boa_engine/src/value/mod.rs

```diff
@@ -88,184 +132,817 @@ impl JsValue {
     #[inline]
-    pub fn nan() -> Self {
-        Self::Rational(f64::NAN)
+    pub fn is_null(&self) -> bool {
+        self.value.get() == NULL_TYPE
     }
 
-    /// Creates a new number with `Infinity` value.
+    /// Returns true if the value is null or undefined.
     #[inline]
-    pub fn positive_infinity() -> Self {
-        Self::Rational(f64::INFINITY)
+    pub fn is_null_or_undefined(&self) -> bool {
+        self.is_null() || self.is_undefined()
     }
 
-    /// Creates a new number with `-Infinity` value.
+    pub fn is_rational(&self) -> bool {
+        (self.value.get() & !SIGN_BIT) <= QNAN
+    }
+
+    fn as_rational_unchecked(&self) -> f64 {
+        f64::from_bits(self.value.get())
+    }
+
+    pub fn as_rational(&self) -> Option<f64> {
+        if self.is_rational() {
+            return Some(self.as_rational_unchecked());
+        }
+
+        None
+    }
+
+    pub fn is_i32(&self) -> bool {
+        self.value.get() & TAG_MASK == INTEGER_TYPE
+    }
+
+    pub fn as_i32_uncheched(&self) -> i32 {
+        (self.value.get() & MASK_INT_PAYLOAD) as u32 as i32
+    }
+
+    pub fn as_i32(&self) -> Option<i32> {
+        if self.is_i32() {
+            return Some(self.as_i32_uncheched());
+        }
+
+        None
+    }
+
+    /// Returns true if the value is a boolean.
     #[inline]
-    pub fn negative_infinity() -> Self {
-        Self::Rational(f64::NEG_INFINITY)
+    pub fn is_boolean(&self) -> bool {
+        self.value.get() & TAG_MASK == BOOLEAN_TYPE
+    }
+
+    pub fn as_boolean_uncheched(&self) -> bool {
+        (self.value.get() & 0xFF) != 0
+    }
+
+    #[inline]
+    pub fn as_boolean(&self) -> Option<bool> {
+        if self.is_boolean() {
+            return Some(self.as_boolean_uncheched());
+        }
+
+        None
+    }
+
+    pub fn as_pointer(&self) -> *mut () {
+        (self.value.get() & MASK_POINTER_PAYLOAD) as *mut ()
     }
 
     /// Returns true if the value is an object
     #[inline]
     pub fn is_object(&self) -> bool {
-        matches!(self, Self::Object(_))
+        self.value.get() & TAG_MASK == OBJECT_TYPE
+    }
+
+    /// Returns a reference to the boxed [`JsObject`] without checking
+    /// if the tag of `self` is valid.
+    ///
+    /// # Safety
+    ///
+    /// Calling this method with a [`JsValue`] that doesn't box
+    /// a [`JsObject`] is undefined behaviour.
+    pub unsafe fn as_object_unchecked(&self) -> &JsObject {
+        // SAFETY: The safety contract must be upheld by the caller
+        let object = unsafe {
+            JsObject::from_void_ptr(self.as_pointer());
+        };
+
+        // SAFETY: Assuming the above is safe, this is safe because
+        // we currently are manipulating a [`ManuallyDrop`] that is
+        // manually dropped until this [`JsValue`] is dropped,
+        // which implies that the lifetime of `&object` can be extended
+        // to the lifetime of `&self`.
+        unsafe { std::mem::transmute(&object) }
```

Comment:
This is UB, we are returning a reference to a local function variable, which can be overridden.

Edit: Same for other `as_` methods.